src/app-layer-expectation.c

```diff
@@ -360,6 +360,38 @@ AppProto AppLayerExpectationHandle(Flow *f, int direction)
     return alproto;
 }
 
+
```

Comment:
I don't understand what the git message says about delay of flow destruction and how it relates to expectations.

Flows can disappear quickly if we're under resource stress, so what happens if the flow is evicted before (or during) the use of the expectation?

Reply:
Regarding the locking, we have 3 cases at expectation check:
- Flow is still there
- Flow has disappeared
- The corner case where expectation is fulfilled at Flow destruction

First case is trivial. Second case too as the Expectation will have been cleaned. The third case is the one we need to care about. The expectation handling is done AppLayerExpectationHandle that locks the IPPair. The expectation destruction in Flow cleaning also lock the IPPair so we have a protection.

Regarding the accuracy of this system, the expectation will be honored if the Flow has not been cleaned. In most protocol, the main flow stays open and thus we don't have any miss. In the unlikely case where a flow is closed, we have a delay of a few seconds between the detection of dead flow and the actual removal. This is plenty of time for most cases as the expectation are usually honored in sub seconds range.